### ISSUE TITLE
Hang when switching language or game data

### ISSUE THREAD

**Issue by AnClark** [k4zmu2a/SpaceCadetPinball#167]
Hi Muzychenko!

When I'm switching language or game data via "Option" on the menu bar, the game hangs (a.k.a. no response).

This is a piece of terminal output when it hangs:

```
(process:19001): GLib-GObject-WARNING **: 22:27:12.935: cannot register existing type 'IpatchSplitsType'

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.935: g_param_spec_enum: assertion 'G_TYPE_IS_ENUM (enum_type)' failed

** (process:19001): CRITICAL **: 22:27:12.936: ipatch_type_install_property: assertion 'G_IS_PARAM_SPEC(prop_spec)' failed

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_boxed_type_register_static: assertion 'g_type_from_name (name) == 0' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchSF2GenType'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot retrieve class for invalid (unclassed) type '<invalid>'

** (process:19001): CRITICAL **: 22:27:12.936: file /build/libinstpatch/src/libinstpatch-1.1.6/libinstpatch/IpatchSF2Gen.c: line 152 (_ipatch_sf2_gen_init): assertion `enum_class != NULL' failed.

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_class_unref: assertion 'g_class != NULL' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchSample'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchItem'

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_interface_add_prerequisite: assertion 'G_TYPE_IS_INTERFACE (interface_type)' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchSF2GenItem'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchItem'

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_interface_add_prerequisite: assertion 'G_TYPE_IS_INTERFACE (interface_type)' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchSF2ModItem'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchItem'

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_interface_add_prerequisite: assertion 'G_TYPE_IS_INTERFACE (interface_type)' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchSF2VoiceCache'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot retrieve class for invalid (unclassed) type '<invalid>'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchItem'

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_register_static: assertion 'parent_type > 0' failed

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_register_static: assertion 'parent_type > 0' failed

(process:19001): GLib-CRITICAL **: 22:27:12.936: g_once_init_leave: assertion 'result != 0' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot retrieve class for invalid (unclassed) type '<invalid>'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchItem'

(process:19001): GLib-GObject-CRITICAL **: 22:27:12.936: g_type_register_static: assertion 'parent_type > 0' failed

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot retrieve class for invalid (unclassed) type '<invalid>'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot register existing type 'IpatchConverter'

(process:19001): GLib-GObject-WARNING **: 22:27:12.936: cannot retrieve class for invalid (unclassed) type '<invalid>'

```

**Comment by k4zmu2a** [comment]
Hello.
I could not reproduce this with the latest game code using Debian Sid and Ubuntu.

According to the logs, the problem happens in libinstpatch, which is related to fluidsynth.
SDL mixer uses fluidsynth for midi playback.
Since the bug only happens on partial restarts, it is probably related to this SDL bug: https://www.fluidsynth.org/news/2021/01/23/sdl2-mixer-bug/.
Try updating your fluidsynth and sdl2_mixer.

**Comment by AnClark** [comment]
> Try updating your fluidsynth and sdl2_mixer.

Thanks! I'll give it a try.

BTW, I'm using Arch Linux, and I install Space Cadet via AUR (`spacecadetpinball-git`).

**Comment by AnClark** [comment]
Problem still exists on Arch Linux, with the latest components:

- Fluidsynth 2.3.0
- SDL2 Mixer 2.6.2
- SDL2 2.26.0

**Comment by k4zmu2a** [comment]
You have the latest SDL libs, so your issue could not be related to that old SDL bug.
I don’t have any other ideas why the game freezes on Arch and not on the other distros.
It could be that Arch version of SDL is somehow different, maybe patched.

Try building the game from this older commit: https://github.com/k4zmu2a/SpaceCadetPinball/commit/3c6e1c9d47445d0f27cf2044dd02b095b5bbba5d.
I recently changed table restart procedure, maybe that somehow broke the game on Arch.

**Comment by k4zmu2a** [comment]
Ok, it looks like this audio reload freeze on Arch is not related to recent table reload changes.

I made some changes to the audio handling in partial table restart: https://github.com/k4zmu2a/SpaceCadetPinball/commit/2162cac9771bb50059f440c50e9a30d982cdd848.
If my guess is correct, on Arch, this version should be able to do partial restarts and only hang on exit.

**Comment by AnClark** [comment]
Sorry for late reply because of a bulk of work. I'll test on my Arch Linux machine.

**Comment by k4zmu2a** [comment]
Released in V2.1.0
